### PR TITLE
Fixes #18006: pass sigma to "Guarded"'s printer

### LINKED ISSUE #18006: Anomaly "Uncaught exception Not_found." in exception printer with Guarded
#### Description of the problem
Error: <in exception printer>: Anomaly "Uncaught exception Not_found."
      Please report at http://coq.inria.fr/bugs/.
      <original exception>: Uncaught exception Type_errors.TypeError(_, _).

#### Coq Version
mukeshtiwari@Mukeshs-MacBook-Pro Semiring_graph_algorithm % coqtop -v
The Coq Proof Assistant, version 8.17.1
compiled with OCaml 5.0.0

The following Coq code triggered the uncaught exception. 
```
Require Import Coq.Lists.List.
Require Import Coq.Relations.Relation_Operators.
Require Import Coq.Wellfounded.Lexicographic_Product.
Require Import Coq.Arith.Wf_nat.

Import ListNotations.

Require Import Coq.Init.Nat.
Require Import Lia.

Require Import Recdef.
Require Import FunInd.
Require Import Coq.Program.Wf.

Declare Scope regex_scope.
Open Scope regex_scope.

Section Regex.

Variable (A : Type).

Inductive Regex : Type :=
| Epsilon : Regex
| CharClass : (A -> bool) -> Regex
| Concat : Regex -> Regex -> Regex
| Union : Regex -> Regex -> Regex
| Star : Regex -> Regex
.

Inductive parse_tree : Type :=
| parse_epsilon : parse_tree
| parse_charclass : parse_tree
| parse_concat : parse_tree -> parse_tree -> parse_tree
| parse_union_l : parse_tree -> parse_tree
| parse_union_r : parse_tree -> parse_tree
| parse_star_nil : parse_tree
| parse_star_cons : parse_tree -> parse_tree -> parse_tree
.

Fixpoint parse_length (t : parse_tree) : nat :=
  match t with
  | parse_epsilon => 0
  | parse_charclass => 1
  | parse_concat t1 t2 => parse_length t1 + parse_length t2
  | parse_union_l t => parse_length t
  | parse_union_r t => parse_length t
  | parse_star_nil => 0
  | parse_star_cons t1 t2 => parse_length t1 + parse_length t2
end.


Definition ε := Epsilon.
Infix "·" := Concat (at level 60, right associativity) : regex_scope.
Infix "∪" := Union (at level 50, left associativity) : regex_scope.
Notation "r *" := (Star r) (at level 40, no associativity) : regex_scope.

Inductive regex_struct : Regex -> Regex -> Prop :=
| regex_struct_concat_l : forall (r1 r2 : Regex),
    regex_struct r1 (r1 · r2)
| regex_struct_concat_r : forall (r1 r2 : Regex),
    regex_struct r2 (r1 · r2)
| regex_struct_union_l : forall (r1 r2 : Regex),
    regex_struct r1 (r1 ∪ r2)
| regex_struct_union_r : forall (r1 r2 : Regex),
    regex_struct r2 (r1 ∪ r2)
| regex_struct_star : forall (r : Regex),
    regex_struct r (r *)
.


Notation "x >>= f" := (flat_map f x) (at level 42, left associativity).

Definition all_parse_trees : 
  list A -> nat ->  Regex * nat -> list parse_tree.
Proof.
  intros w start rn.
  cut (Acc (slexprod _ _ regex_struct lt) rn).
  intros Hacc.
  revert start w. (* arugments are chaned a bit *)
  revert dependent rn.
  refine
    (fix Fn rn Hacc {struct Hacc} := _).
  refine 
    (match rn as rn' return rn = rn' -> _ with 
    | (Epsilon, _) => fun _ _ _  => [parse_epsilon]
    | (CharClass p, _) => fun Ha start w => 
      match nth_error w start with
      | Some a => if p a then [parse_charclass] else []
      | None => []
      end
    | (r1 · r2, len) => fun Ha start w => 
      Fn (r1, len) _ start w >>= fun t1 =>
        map (fun t2 => parse_concat t1 t2) 
        (Fn (r2, (len - parse_length t1)) _ 
        (start + parse_length t1) w)
    | (r1 ∪ r2, len) => fun Ha start w => 
      map (fun t => parse_union_l t) (Fn (r1, len) _ start w) ++
      map (fun t => parse_union_r t) (Fn (r2, len) _ start w)
    | ((r *), len) => fun Ha start w => 
      match len with
      | 0 => [parse_star_nil]
      | _ => 
      (Fn (r, len) _ start w >>= fun t1 =>
        match parse_length t1 with 
        | 0 => []
        | l1 => map (fun t2 => parse_star_cons t1 t2)  
          (Fn (r *, (len - l1)) _ (start + parse_length t1) w)
          ++ [parse_star_nil] end)
      end
    end eq_refl); subst.
    Guarded.
    (* The condition holds up to here *) 
    +
      eapply wf_slexprod.
      Guarded.
      (* 
      Error: <in exception printer>: Anomaly "Uncaught exception Not_found."
      Please report at http://coq.inria.fr/bugs/.
      <original exception>: Uncaught exception Type_errors.TypeError(_, _).
      *)
    
```

### PR DESCRIPTION
**Pull request by herbelin**
Fixes #18006

- [x] Added / updated **test-suite**.
- [x] Added **changelog**.

### CODE REVIEW

**Review by ppedrot — APPROVED** [review]

**Comment by ppedrot** [comment]
@coqbot run full ci

File: pretyping/inductiveops.ml [hunk 1]
```diff
@@ -716,4 +716,6 @@ let control_only_guard env sigma c =
     check_fix_cofix env c;
     EConstr.iter_with_full_binders env sigma EConstr.push_rel iter env c
   in
-  iter env c
+  try iter env c
+  with Type_errors.TypeError (env, e) ->
+    Loc.raise (Pretype_errors.PretypeError (env, sigma, TypingError (Type_errors.map_ptype_error EConstr.of_constr e)))
```
Comment (SkySkimmer) [inline pretyping/inductiveops.ml:721]: Why Loc.raise? IMO should be either Exninfo.raise with an info captured from the kernel exception, or unqualified raise

Comment (herbelin) [reply pretyping/inductiveops.ml:721]: > Why Loc.raise?

I don't know. I just copied what was done for `Pretyping.esearch_guard`.

I'd be ok with `Exninfo.raise` too. Should we then change also `Pretyping.esearch_guard`?

Comment (ppedrot) [reply pretyping/inductiveops.ml:721]: Unqualified raise is probably better, we don't really care about the backtrace at this point.

Comment (ppedrot) [reply pretyping/inductiveops.ml:721]: Yes.

**Comment by ppedrot** [comment]
@coqbot merge now

**Pull request merged**